python/lsst/afw/cameraGeom/camera.py

```diff
@@ -68,118 +73,73 @@ def getPupilFactory(self, visitInfo, pupilSize, npix, **kwargs):
-        if isinstance(toSys, CameraSysPrefix):
-            # Must be in a detector.  Find the detector and transform it.
-            detList = self.findDetectors(nativePoint)
-            if len(detList) == 0:
-                raise RuntimeError("No detectors found")
-            elif len(detList) > 1:
-                raise RuntimeError("More than one detector found")
-            else:
-                det = detList[0]
-                return det.transform(nativePoint, toSys)
-        else:
-            return self._transformSingleSys(nativePoint, toSys)
+        `fromSys` and `toSys` must both be present in the same transform map,
+        but that transform map may be from any detector or this camera object.
 
-    def _transformSingleSysArray(self, positionList, fromSys, toSys):
-        """!Transform an array of points from once CameraSys to another CameraSys
-        @warning This method only handles a single jump, not a transform linked by a common native sys.
+        @param[in] fromSys  Camera coordinate system of `position`
+                        input points
+        @param[in] toSys  Camera coordinate system of returned point(s)
 
-        @param[in] positionList List of Point2D objects, one per position
-        @param[in] fromSys  Initial camera coordinate system
-        @param[in] toSys  Destination camera coordinate system
+        @return an lsst.afw.geom.TransformPoint2ToPoint2 that transforms from
+            `fromSys` to `toSys` in the forward direction
 
-        @returns an array of Point2D objects containing the transformed coordinates in the destination system.
+        @throws lsst.pex.exceptions.InvalidParameter if no transform is
+        available. This includes the case that fromSys specifies a known
+        detector and toSys specifies any other detector (known or unknown)
+        @throws KeyError if an unknown detector is specified
         """
         if fromSys.hasDetectorName():
-            det = self[fromSys.getDetectorname()]
-            detTrans = det.getTransfromMap()
-            return detTrans.transform(positionList, fromSys, toSys)
+            det = self[fromSys.getDetectorName()]
+            return det.getTransformMap().getTransform(fromSys, toSys)
         elif toSys.hasDetectorName():
             det = self[toSys.getDetectorName()]
-            detTrans = det.getTransformMap()
-            return detTrans.transform(positionList, fromSys, toSys)
-        elif toSys in self._transformMap:
-            # use camera transform map
-            return self._transformMap.transform(positionList, fromSys, toSys)
-        raise RuntimeError(
-            "Could not find mapping from %s to %s"%(fromSys, toSys))
-
-    def _transformSingleSys(self, cameraPoint, toSys):
-        """!Transform a CameraPoint with a CameraSys to another CameraSys.
+            return det.getTransformMap().getTransform(fromSys, toSys)
+        else:
+            return self.getTransformMap().getTransform(fromSys, toSys)
```

Comment:
Your suggested new code does not do the same thing because the `else` clause handles the camera's own transform map, not a detector's transform map.

I suppose we could retrieve the transform map. It will be a bit awkward since we then have to assume the code will never leave the transform map equal to None (which will be true initially, but perhaps not if the code evolves) or check for None even though it should be impossible. That is why I chose to use the current technique.